secio/src/dh_compat/mod.rs

```diff
@@ -17,6 +17,56 @@ pub enum KeyAgreement {
     X25519,
 }
 
+#[cfg(all(ossl110, unix))]
+pub use openssl_impl::*;
+#[cfg(not(target_arch = "wasm32"))]
+#[cfg(not(unix))]
+pub use ring_impl::*;
+#[cfg(all(not(ossl110), unix))]
+pub use ring_openssl_unix::*;
+
+#[cfg(all(not(ossl110), unix))]
+mod ring_openssl_unix {
+    use super::KeyAgreement;
+    use super::{openssl_impl, ring_impl};
+    use crate::error::SecioError;
+
+    pub enum EphemeralPrivateKey {
+        Openssl(openssl_impl::EphemeralPrivateKey),
+        Ring(ring_impl::EphemeralPrivateKey),
+    }
+
+    pub fn generate_agreement(
+        algorithm: KeyAgreement,
+    ) -> Result<(EphemeralPrivateKey, Vec<u8>), SecioError> {
+        match algorithm {
+            KeyAgreement::EcdhP256 | KeyAgreement::EcdhP384 => {
+                openssl_impl::generate_agreement(algorithm)
+                    .map(|(private, pubkey)| (EphemeralPrivateKey::Openssl(private), pubkey))
+            }
+            KeyAgreement::X25519 => ring_impl::generate_agreement(algorithm)
+                .map(|(private, pubkey)| (EphemeralPrivateKey::Ring(private), pubkey)),
+        }
+    }
+
+    pub fn agree(
+        algorithm: KeyAgreement,
+        my_private_key: EphemeralPrivateKey,
+        other_public_key: &[u8],
+    ) -> Result<Vec<u8>, SecioError> {
+        match (algorithm, my_private_key) {
+            (
+                KeyAgreement::EcdhP256 | KeyAgreement::EcdhP384,
+                EphemeralPrivateKey::Openssl(private_key),
+            ) => openssl_impl::agree(algorithm, private_key, other_public_key),
+            (KeyAgreement::X25519, EphemeralPrivateKey::Ring(private_key)) => {
+                ring_impl::agree(algorithm, private_key, other_public_key)
+            }
+            _ => unreachable!(),
```

Comment:
`panic!()` here would be better? In case user with incorrect combination of `KeyAgreement` and `EphemeralPrivateKey`.

Reply:
In our code, this branch is unreachable and will panic here